### PR TITLE
Change NetworkCurl according to os_adapation::NetworkProtocolCurl.

### PR DESCRIPTION
**Pull request by LiubovDidkivska**
Functions Run, CompleteMessage, Initialize, Deinitialize are changed.
LostHandles functionality removed from Run function.

Relates-To:OLPEDGE-1694

Signed-off-by: Liubov Didkivska <ext-liubov.didkivska@here.com>

### CODE REVIEW

File: olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp [hunk 1]
```diff
@@ -871,59 +850,52 @@ size_t NetworkCurl::HeaderFunction(char* ptr, size_t size, size_t nitems,
...
+    RequestHandle& rhandle = handles_[index];
 
-  std::vector<std::pair<std::string, std::string> > statistics;
-  if (index < handles_.size()) {
-    if (handles_[index].get_statistics) {
-      statistics =
-          GetStatistics(handles_[index].handle, handles_[index].retry_count);
+    std::vector<std::pair<std::string, std::string> > statistics;
+    if (rhandle.get_statistics) {
+      statistics = GetStatistics(rhandle.handle, rhandle.retry_count);
```
Comment (andescu) [inline olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp]: Are these even used?

File: olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp [hunk 2]
```diff
@@ -871,59 +850,52 @@ size_t NetworkCurl::HeaderFunction(char* ptr, size_t size, size_t nitems,
...
     std::string error("Success");
     int status;
     if ((result == CURLE_OK) || (result == CURLE_HTTP_RETURNED_ERROR)) {
       long http_status = 0;
-      curl_easy_getinfo(handles_[index].handle, CURLINFO_RESPONSE_CODE,
-                        &http_status);
+      curl_easy_getinfo(rhandle.handle, CURLINFO_RESPONSE_CODE, &http_status);
       status = static_cast<int>(http_status);
-      if ((handles_[index].offset == 0) && (status == 206)) status = 200;
+      if ((rhandle.offset == 0) && (status == 206))
```
Comment (andescu) [inline olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp]: Always add braces please.

File: olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp [hunk 3]
```diff
@@ -871,59 +850,52 @@ size_t NetworkCurl::HeaderFunction(char* ptr, size_t size, size_t nitems,
...
   std::unique_lock<std::mutex> lock(event_mutex_);
-  size_t index = 0;
-  for (; index < handles_.size(); ++index) {
-    if (handles_[index].in_use && (handles_[index].handle == handle)) {
-      break;
-    }
-  }
+  int index = GetHandleIndex(handle);
+  if (index >= 0 && index < static_cast<int>(handles_.size())) {
+    RequestHandle& rhandle = handles_[index];
```
Comment (andescu) [inline olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp]: Usually, abbreviations are not allowed but we can close one eye for this.
Else i would suggest to name it `request_handle` or just `handle`.

File: olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp [hunk 4]
```diff
@@ -1008,17 +979,38 @@ void NetworkCurl::Run() {
...
+                }
               }
             }
             break;
           }
           case EventInfo::Type::CANCEL_EVENT:
             if (event.handle->in_use) {
-              curl_multi_remove_handle(curl_, event.handle->handle);
-              event_mutex_.unlock();
+              CURLMcode code =
```
Comment (andescu) [inline olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp]:
```suggestion
              auto code =
```

File: olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp [hunk 5]
```diff
@@ -1087,184 +1080,63 @@ void NetworkCurl::Run() {
...
-      maxfd = pipe_[0];
-    }
+    {
+      int numfds = 0;
+#if defined(OLP_SDK_NETWORK_HAS_PIPE) || defined(OLP_SDK_NETWORK_HAS_PIPE2)
+      curl_waitfd waitfd[1];
+      waitfd[0].fd = pipe_[0];
+      waitfd[0].events = CURL_WAIT_POLLIN;
+      waitfd[0].revents = 0;
+      CURLMcode mc = curl_multi_wait(curl_, waitfd, 1, 1000, &numfds);
```
Comment (andescu) [inline olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp]:
```suggestion
      auto mc = curl_multi_wait(curl_, waitfd, 1, 1000, &numfds);
```

File: olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp [hunk 6]
```diff
@@ -1087,184 +1080,63 @@ void NetworkCurl::Run() {
...
+      // read pipe data if it is signaled
+      if (mc == CURLM_OK && numfds != 0 && waitfd[0].revents != 0) {
+        char tmp;
+        while (read(waitfd[0].fd, &tmp, 1) > 0)
+          ;
+      }
+#else
+      // Without pipe limit wait time to 100ms
+      // so that network events can be handled in a reasonable time
+      CURLMcode mc = curl_multi_wait(curl_, nullptr, 0, 100, &numfds);
```
Comment (andescu) [inline olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp]:
```suggestion
      auto mc = curl_multi_wait(curl_, nullptr, 0, 100, &numfds);
```

File: olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp [hunk 7]
```diff
@@ -1087,184 +1080,63 @@ void NetworkCurl::Run() {
...
-              ReleaseHandle(&handles_[handle_index]);
-            }
-          }
-        }
-      }
-      if (!IsStarted()) {
-        continue;
-      }
-      bool in_use_handles = false;
+        bool inUseHandles = std::any_of(
```
Comment (andescu) [inline olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp]:
```suggestion
        bool in_use_handles = std::any_of(
```

File: olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp [hunk 8]
```diff
@@ -1087,184 +1080,63 @@ void NetworkCurl::Run() {
...
 
-      // TODO: examine this section in details. Not clear what we are waiting
-      // for.
-      if (timeout < 0) {
-        if (!in_use_handles) {
+        if (!inUseHandles) {
           // Enter wait only when all handles are free
           event_condition_.wait_for(lock, std::chrono::seconds(2));
         } else {
+          const int WAIT_MSEC = 100;
```
Comment (andescu) [inline olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp]:
```suggestion
          constexpr int kWaitMsec = 100;
```

File: olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp [hunk 9]
```diff
@@ -237,13 +235,7 @@ NetworkCurl::NetworkCurl(size_t max_requests_count)
 
 NetworkCurl::~NetworkCurl() {
   OLP_SDK_LOG_TRACE(kLogTag, "Destroyed NetworkCurl object, this=" << this);
-
-  if (state_ == WorkerState::STARTED) {
     Deinitialize();
-  }
-  if (curl_initialized_) {
-    curl_global_cleanup();
```
[the remark this answers is not in the document]
Comment (LiubovDidkivska) [reply olp-cpp-sdk-core/src/http/curl/NetworkCurl.cpp]: Here curl_global_init was added, but not curl_global_cleanup. But I agree it makes sense to have  curl_global_cleanup https://gerrit.it.here.com/gitweb?p=os-adaptation/network.git;a=commitdiff;h=d8619c321b45b5d794058e85e1559ba43f75727f

**Pull request merged**